### PR TITLE
Support multiple course orgs per Site

### PR DESCRIPTION
**Pull request by bryanlandia**
Corollary to edx-platform PR https://github.com/appsembler/edx-platform/pull/785.  Don't fail if SiteConfiguration has multiple values for `course_org_filter`

To be clear, we are not going to enable this feature flag on Tahoe multitenant, it's only for standalone customer sites.

### CODE REVIEW

**Comment by bryanlandia** [comment]
This just makes the auth check not fail—that's it.  It doesn't magically make Tahoe's organizations-based setup work for multiple course orgs.  It does get us a big part of what we need for Tahoe standalone with multiple course orgs though.

**Review by OmarIthawi — CHANGES_REQUESTED** [review]
Thanks Bryan.

Can we please use the same feature flag and add tests for it? It would be a hack, but more well formed that this one in my opinion.

Also, how can Figures support this new change? If it's a blocker, we might want to consider a another approach since it's unclear to me why we're doing this.

Is it for URL purposes? e.g. we don't want to be able to change `course-v1:ORG`? or it's about permissions or both?

**Comment by bryanlandia** [comment]
Thanks @OmarIthawi .  Yes, I can set up with a FEATURE flag and add tests.  I notice there are no tests for the auth backend at all, so far.  

It's not exactly for URL purposes.  For example, HISP/DHIS2 uses 6 or 7 different orgs to reflect the various partners with which they offer courses.  PSU does the same (though they are still on Ginkgo).  We actually key some customizations for certificates off of those course orgs.  We may be able to consolidate toward a single course org in the future, and then use display org strings, but we can't do that for currently running courses.

**Comment by OmarIthawi** [comment]
@bryanlandia thanks for the explaination. So I see some missing product/customer needs in Tahoe. For such feedback I would love if you write them in a confluence page and pass them to Matej.

It is important if such changes affects core repositories as opposed to plugins or XBlocks.

**Comment by bryanlandia** [comment]
@OmarIthawi I added a set of tests for previous functionality of OrganizationMemberBackend as well as to exercise the new FEATURE.  I found and fixed some edge cases in case of a blank course_org_filter value.  Are we using and checking that value for performance reasons, rather than just looking at the comparison between user.organizations and site.organizations?

cc @melvinsoft

**Review by OmarIthawi — DISMISSED** [review]
Thanks @bryanlandia. I have to put this PR on hold until @johnbaldwin can review it. Figures could block this.

Sorry that it's coming at the last time, but we noticed that Figures complexity is increasing and I'm advocating for keeping the system (hence Figures) as simple as possible.

Anyway, let's hear from John and see how can we make it work.

**Comment by bryanlandia** [comment]
Thanks @omar.  @johnbaldwin it would be great to get your review soon.  I'll do some checking in Figures as well

**Comment by bryanlandia** [comment]
I added to the PR description

> To be clear, we are not going to enable this feature flag on Tahoe multitenant, it's only for standalone customer sites.

File: organizations/backends.py [hunk 1]
```diff
@@ -34,11 +50,37 @@ def authenticate(self, *args, **kwargs):
...
+            # Site must have a defined course org.  Treat a blank value as None
+            if site_organizations is None or len(site_organizations) == 0:
+                return None
+
+            # get a sequence and remove dupes
+            if not isinstance(site_organizations, (tuple, collections.MutableSequence, type(None))):
+                site_organizations = (site_organizations, )
+            site_organizations = [org for org in set(site_organizations) if len(org)]
+
+            try:
```
Comment (johnbaldwin) [inline organizations/backends.py:67]: Why use exception handling and an assert here instead of

```
if not len(site_organizations) == 1:
    logger....
    return None
```

Also, this appears to be an error case due to some kind of misconfiguration. Shouldn't the logger be at least error level?

File: organizations/backends.py [hunk 2]
```diff
@@ -1,8 +1,24 @@
+import collections
+import logging
+
 from django.conf import settings
 from django.contrib.auth.backends import ModelBackend
 from django.core.exceptions import PermissionDenied
-from openedx.core.djangoapps.theming.helpers import get_current_site
-from openedx.core.djangoapps.site_configuration.helpers import is_site_configuration_enabled, get_value
+
+try:
```
Comment (johnbaldwin) [inline organizations/backends.py:8]: @bryanlandia First, thanks for adding test coverage for backends, much needed addition!

However I strongly urge against embedding test code in production code.  I understand the challenge of mocking imports to modules that don't exist for the test environment and this was the solution you found. However, I see it as an avoidable anti-pattern. 

I had to address the same problem in `figures.compat` in order to test agains non-existing platform code.  Here is how I did it:

The basic approach is to patch sys.modules for your test using `mock.patch.dict`. Here is a convenience function I created:

https://github.com/appsembler/figures/blob/master/tests/test_compat.py#L33-L60

Then in the test, you can simply use `patch_module` as following: 

https://github.com/appsembler/figures/blob/master/tests/test_compat.py#L130-L161

@OmarIthawi What are your thoughts here?

Comment (OmarIthawi) [reply organizations/backends.py:8]: Regarding mocking, I think Figures code provides a lot of quality examples which I've reused in the [Course Access Groups repo](https://github.com/appsembler/course-access-groups).

File: organizations/backends.py [hunk 3]
```diff
@@ -1,8 +1,24 @@
...
+    from openedx.core.djangoapps.theming.helpers import get_current_site
+    from openedx.core.djangoapps.site_configuration.helpers import is_site_configuration_enabled, get_value
+except ImportError:
+    # we mock them out in test as we don't have edx-platform
+    from os import getenv
+    if 'test' in getenv('DJANGO_SETTINGS_MODULE'):
+        # set these so they are mockable from tests
+        def get_current_site(): pass
+        def get_value(_): pass
+        def is_site_configuration_enabled(): pass
```
Comment (OmarIthawi) [inline organizations/backends.py:18]: I agree with John here this should be mocked by tests instead of being in production code and potentially breaking the code silently.

**Review by OmarIthawi — COMMENTED** [review]
Thanks again @bryanlandia for waiting so much on reviews. The first couple of PRs are the hardest to get since they provide the biggest changes for Tahoe.

For this PR, I would like to take another look it and related PRs, and review what [Bryan have wrote so far in the shared DHIS2 document](https://appsembler.atlassian.net/wiki/spaces/ED/pages/1302364509/Making+Tahoe+work+for+DHIS2#Changes-to-be-made-on-Tahoe-for-DHIS2%2FStandalone-Unification).

Do you have any timeline for this PR that you'd like to get it concluded by?

**Comment by johnbaldwin** [comment]
@bryanlandia @OmarIthawi Why do we use `openedx.core.djangoapps.theming.helpers.get_current_site` instead of `django.contrib.sites.shortcuts.get_current_site`

File: organizations/backends.py [hunk 4]
```diff
@@ -34,11 +50,37 @@ def authenticate(self, *args, **kwargs):
         # superuser can log into any microsite
         site = get_current_site()
         is_default_site = site.id == settings.SITE_ID
+
         if not is_default_site and is_site_configuration_enabled() and user and not user.is_superuser:
```
Comment (johnbaldwin) [inline organizations/backends.py:54]: @OmarIthawi or @bryanlandia I know this is code not modified in this PR and also this line here is original code from Filip, but do you know why we need to make the call to `is_site_configuration_enabled{}` ? 

When would a site NOT have a "configuration" attribute?

Here's the source and "blame" for the `is_site_configuration_enabled` function:
The code is from upstream: https://github.com/appsembler/edx-platform/blame/appsembler/tahoe/master/openedx/core/djangoapps/site_configuration/helpers.py#L30-L41

Why could we not check if there is an `organizations.models.UserOrganizationMapping` model instead?

**Comment by johnbaldwin** [comment]
@bryanlandia @OmarIthawi Another question:

Where will our fork / this branch of `organizations` be run? Can we assume that are mapping organizations to sites and users to organziations for every deployment using this branch of our fork?

File: organizations/backends.py [hunk 5]
```diff
@@ -34,11 +50,37 @@ def authenticate(self, *args, **kwargs):
...
             user_organizations = set(user.organizations.values_list('short_name', flat=True))
-            site_organization = get_value('course_org_filter')
```
Comment (johnbaldwin) [inline organizations/backends.py:39]: @OmarIthawi @bryanlandia Any idea why we're doing this instead of simply checking

```
if UserOrganizationMapping.objects.filter(user_id=user.id, organization__sites__in=[site]).exists():
    do_our_stuff()
```

File: organizations/backends.py (hunk 4 above)
Comment (OmarIthawi) [reply organizations/backends.py:54]: I don't really know.

File: organizations/backends.py (hunk 5 above)
Comment (bryanlandia) [reply organizations/backends.py:39]: I believe we have followed this pattern of using `course_org_filter` because it is more performant but I will give it another think.

**Comment by thraxil** [comment]
Is this still being worked on?

**Comment by OmarIthawi** [comment]
@bryanlandia is this being used anywhere?

We're refactoring all changes of this repository into https://github.com/appsembler/tahoe-sites/ , so this pull request should be either:

 - Migrated into https://github.com/appsembler/tahoe-sites/ behind a feature flag (including tests)
 - Migrated into a forked https://github.com/appsembler/tahoe-sites/ branch e.g. `my-customer/juniper` branch (can be done to save on testing cost)
 - Closed and added as a Jira task in Black Team for future follow up
 
We're closing this repository and marking it as archive after the migration is complete.

What do you think?

**Comment by thraxil** [comment]
Is this still being worked on?

**Comment by OmarIthawi** [comment]
Thanks Bryan for the work here. I know this we'll probably need to create an equivalent solution but this is opened against Hawthorn and `tahoe-sites` needs to be refactored in any new solution. So,  I'm going to close this PR.

Thanks Anders for the reminder.

**Comment by bryanlandia** [comment]
No problem, was fine to close.
